src/analysis/convex/cone.lean

```diff
@@ -185,6 +186,153 @@ ext' $ preimage_comp.symm
 @[simp] lemma mem_comap {f : E →ₗ[ℝ] F} {S : convex_cone F} {x : E} :
   x ∈ S.comap f ↔ f x ∈ S := iff.rfl
 
+/-! ### Convex cones with extra properties -/
+
+/-- A convex cone is pointed if it includes 0. -/
+def pointed (S : convex_cone E) : Prop := (0 : E) ∈ S
+
+/-- A convex cone is blunt if it doesn't include 0. -/
+def blunt (S : convex_cone E) : Prop := (0 : E) ∉ S
+
+/-- A convex cone is flat if it contains some nonzero vector `x` and its opposite `-x`. -/
+def flat (S : convex_cone E) : Prop := ∃ x ∈ S, x ≠ (0 : E) ∧ -x ∈ S
+
+/-- A convex cone is salient if it doesn't include `x` and `-x` for any nonzero `x`. -/
+def salient (S : convex_cone E) : Prop := ∀ x ∈ S, x ≠ (0 : E) → -x ∉ S
+
+lemma pointed_iff_not_blunt (S : convex_cone E) : pointed S ↔ ¬blunt S :=
+⟨λ h₁ h₂, h₂ h₁, λ h, not_not.mp h⟩
+
+lemma salient_iff_not_flat (S : convex_cone E) : salient S ↔ ¬flat S :=
+begin
+  split,
+  { rintros h₁ ⟨x, xs, H₁, H₂⟩,
+    exact h₁ x xs H₁ H₂ },
+  { intro h,
+    unfold flat at h,
+    push_neg at h,
+    exact h }
+end
+
+/-- A blunt cone (one not containing 0) is always salient. -/
+lemma salient_of_blunt (S : convex_cone E) : blunt S → salient S :=
+begin
+  intro h₁,
+  rw [salient_iff_not_flat],
+  intro h₂,
+  obtain ⟨x, xs, H₁, H₂⟩ := h₂,
+  have hkey : (0 : E) ∈ S := by rw [(show 0 = x + (-x), by simp)]; exact add_mem S xs H₂,
+  exact h₁ hkey,
+end
+
+/-- A pointed convex cone defines a preorder. -/
+def to_preorder (S : convex_cone E) (h₁ : pointed S) : preorder E :=
+{ le := λ x y, y - x ∈ S,
+  le_refl := λ x, by change x - x ∈ S; rw [sub_self x]; exact h₁,
+  le_trans := λ x y z xy zy, by simp [(show z - x = z - y + (y - x), by abel), add_mem S zy xy] }
+
+/-- A pointed and salient cone defines a partial order. -/
+def to_partial_order (S : convex_cone E) (h₁ : pointed S) (h₂ : salient S) : partial_order E :=
+{ le_antisymm :=
+    begin
+      intros a b ab ba,
+      by_contradiction h,
+      have h' : b - a ≠ 0 := λ h'', h (eq_of_sub_eq_zero h'').symm,
+      have H := h₂ (b-a) ab h',
+      rw [neg_sub b a] at H,
+      exact H ba,
+    end,
+  ..to_preorder S h₁ }
+
+/-- A pointed and salient cone defines an `ordered_add_comm_group`. -/
+def to_ordered_add_comm_group (S : convex_cone E) (h₁ : pointed S) (h₂ : salient S) :
+  ordered_add_comm_group E :=
+{ add_le_add_left :=
+    begin
+      intros a b hab c,
+      change c + b - (c + a) ∈ S,
+      rw [add_sub_add_left_eq_sub],
+      exact hab,
+    end,
+  ..to_partial_order S h₁ h₂,
+  ..show add_comm_group E, by apply_instance }
+
+/--
+Constructs an ordered semimodule given an `ordered_add_comm_group`, a cone, and a proof that
+the order relation is the one defined by the cone.
+-/
+def to_ordered_semimodule {M : Type*} [ordered_add_comm_group M] [semimodule ℝ M]
```

Comment:
This doesn't have to do with the "extra properties" mentioned in the sectioning module doc string and doesn't rely on any results there either, so I think it should be moved before, up to line 189.